File: src/anlas.ts
import {
  MAX_NAI_UPSCALE_INPUT_PIXELS,
  type AccountSummary,
  type AnlasQuoteFeature,
  type AnlasQuoteResult,
  type DirectorTool,
  type GenerateExtras,
  type GenerateParams,
  type I2IParams,
  type NAIInpaintModel,
  type NAIModel,
  type UpscaleScale,
  type WorkingImage,
} from "./types";

const BASE_PIXEL_COEFFICIENT = 2951823174884865e-21;
const STEP_PIXEL_COEFFICIENT = 5753298233447344e-22;
const OPUS_FREE_MAX_PIXELS = 1024 * 1024;

function isV4Plus(model: string) {
  return model.includes("-4");
}

function isActiveOpus(account?: AccountSummary) {
  return Boolean(account?.hasActiveSubscription && (account.tierLevel ?? 0) >= 3);
}

function positiveInt(value: unknown, fallback: number) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.max(1, Math.floor(parsed));
}

function clamp01(value: unknown, fallback: number) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.min(1, Math.max(0, parsed));
}

function fitSizeWithinPixels(width: number, height: number, maxPixels: number) {
  const pixels = width * height;
  if (!width || !height || pixels <= maxPixels) return { width, height, resized: false };
  const ratio = Math.sqrt(maxPixels / pixels);
  return {
    width: Math.max(1, Math.floor(width * ratio)),
    height: Math.max(1, Math.floor(height * ratio)),
    resized: true,
  };
}

function finalizeQuote(
  feature: AnlasQuoteFeature,
  amount: number,
  account: AccountSummary | undefined,
  details: string[],
  source: AnlasQuoteResult["source"],
): AnlasQuoteResult {
  const normalized = Math.max(0, Math.ceil(amount));
  const balance = account?.anlasBalance;
  const insufficient = typeof balance === "number" && normalized > balance;
  return {
    ok: true,
    amount: normalized,
    source,
    balance,
    insufficient,
    message: insufficient
      ? `${feature} needs ${normalized} Anlas, but balance is ${balance}.`
      : `${feature} will cost ${normalized} Anlas.`,
    details,
  };
}

export function calculateImageGenerationAnlas({
  params,
  account,
  extras,
  batchCount = 1,
  action = "generate",
  strength = 1,
  forcePaid = false,
  alreadyEncodedVibes = 0,
}: {
  params: GenerateParams;
  account?: AccountSummary;
  extras?: GenerateExtras;
  batchCount?: number;
  action?: "generate" | "img2img" | "infill";
  strength?: number;
  forcePaid?: boolean;
  /** How many of the vibe references are already encoded+cached (no re-charge). */
  alreadyEncodedVibes?: number;
}): AnlasQuoteResult {
  const samples = positiveInt(batchCount, 1);
  const width = positiveInt(params.width, 512);
  const height = positiveInt(params.height, 512);
  const pixels = Math.max(width * height, 65_536);
  const steps = positiveInt(params.steps, 28);
  const normalizedStrength = action === "generate" ? 1 : clamp01(strength, 1);
  const v4Plus = isV4Plus(params.model);
  const vibeCount = extras?.vibeImages?.length ?? 0;
  const details: string[] = [];

  let basePerSample = 0;
  const opusFree =
    !forcePaid &&
    action === "generate" &&
    isActiveOpus(account) &&
    pixels <= OPUS_FREE_MAX_PIXELS &&
    steps <= 28;

  if (opusFree) {
    details.push("Opus active: base text-to-image generation is free for this size/step request.");
  } else {
    const smeaMultiplier = !v4Plus && params.smeaDyn ? 1.4 : !v4Plus && params.smea ? 1.2 : 1;
    const officialBase = Math.ceil(BASE_PIXEL_COEFFICIENT * pixels + STEP_PIXEL_COEFFICIENT * pixels * steps);
    // NovelAI caps a single image's base generation cost at 140 Anlas.
    basePerSample = Math.min(140, Math.max(2, Math.ceil(officialBase * smeaMultiplier * normalizedStrength)));
    details.push(`Base image price: ${basePerSample} Anlas each by the official frontend formula.`);
  }

  let total = basePerSample * samples;

  if (v4Plus && vibeCount > 0) {
    // Per the official docs, encoding an image into a vibe is a ONE-TIME fee of
    // 2 Anlas (the encoding is deterministic and cached), so it is NOT multiplied
    // by the batch/request count — and references already encoded+cached this
    // session incur NO further encode charge, so only count the un-encoded ones.
    const toEncode = Math.max(0, vibeCount - Math.min(vibeCount, alreadyEncodedVibes));
    const encodeCost = 2 * toEncode;
    total += encodeCost;
    details.push(
      `V4+ Vibe encoding (one-time): ${toEncode} of ${vibeCount} image(s) x 2 Anlas = ${encodeCost}` +
        `${alreadyEncodedVibes > 0 ? `（${Math.min(vibeCount, alreadyEncodedVibes)} 张已缓存，不再计费）` : ""}.`,
    );
    if (vibeCount > 4) {
      // The >4-vibe surcharge is added "to the generation", i.e. per request.
      const multiVibeCost = 2 * (vibeCount - 4) * samples;
      total += multiVibeCost;
      details.push(`Extra Vibe fee above 4 references: ${multiVibeCost} Anlas.`);
    }
  }

  // Precise / Director references (V4.5): the official docs charge "an additional
  // cost of 5 Anlas to each image generation" — a FLAT 5 per generated image when
  // the feature is used, NOT per reference. So it scales with the request/batch
  // count only, regardless of how many references are attached.
  // Precise Reference is V4.5-only (per the docs), and the main process drops it
  // on other models — so only charge it for V4.5 to avoid over-quoting.
  const preciseCount = extras?.preciseReferences?.length ?? 0;
  if (v4Plus && params.model.includes("4-5") && preciseCount > 0) {
    const preciseCost = 5 * samples;
    total += preciseCost;
    details.push(`Precise reference: 5 Anlas x ${samples} image(s) = ${preciseCost} (flat per image, ${preciseCount} ref attached).`);
  }

  return finalizeQuote(action === "generate" ? "generate" : action === "img2img" ? "i2i" : "inpaint", total, account, details, "estimate-formula");
}

export function calculateUpscaleAnlas({
  image,
  account,
  scale,
}: {
  image?: Pick<WorkingImage, "width" | "height"> | null;
  account?: AccountSummary;
  scale?: UpscaleScale;
}): AnlasQuoteResult {
  if (!image?.width || !image?.height) {
    return { ok: false, source: "unavailable", message: "请先加载要超分的图片，才能读取生成前扣费。" };
  }
  const prepared = fitSizeWithinPixels(image.width, image.height, MAX_NAI_UPSCALE_INPUT_PIXELS);
  const pixels = prepared.width * prepared.height;
  const details = [
    prepared.resized
      ? `Input is pre-shrunk for NovelAI upscale: ${image.width}x${image.height} -> ${prepared.width}x${prepared.height}.`
      : `Input size used for upscale quote: ${prepared.width}x${prepared.height}.`,
    `Upscale scale: ${scale ?? 4}x.`,
  ];
  if (isActiveOpus(account) && pixels <= 409_600) {
    details.push("Opus active: official upscale tier is free for this input size.");
    return finalizeQuote("upscale", 0, account, details, "estimate-formula");
  }
  let amount = -3;
  if (pixels <= 262_144) amount = 1;
  else if (pixels <= 409_600) amount = 2;
  else if (pixels <= 524_288) amount = 3;
  else if (pixels <= 786_432) amount = 5;
  else if (pixels <= 1_048_576) amount = 7;
  if (amount < 0) {
    return {
      ok: false,
      source: "unavailable",
      balance: account?.anlasBalance,
      message: "图片分辨率超过 NovelAI 云端超分的报价范围。",
      details,
    };
  }
  return finalizeQuote("upscale", amount, account, details, "estimate-formula");
}

export function calculateDirectorAnlas({
  tool,
  account,
}: {
  tool?: DirectorTool;
  account?: AccountSummary;
}): AnlasQuoteResult {
  const amount = tool === "bg-removal" ? 65 : 0;
  const details =
    tool === "bg-removal"
      ? ["Background removal is a fixed 65 Anlas director-tool request."]
      : ["This director tool is currently free in NovelAI's director-tool pricing."];
  return finalizeQuote("director", amount, account, details, "estimate-fixed");
}

export function calculateFeatureAnlasQuote({
  feature,
  params,
  extras,
  batchCount,
  i2iParams,
  inpaintModel,
  inpaintStrength,
  account,
  image,
  upscaleScale,
  directorTool,
  alreadyEncodedVibes = 0,
}: {
  feature: AnlasQuoteFeature;
  params?: GenerateParams;
  extras?: GenerateExtras;
  batchCount?: number;
  i2iParams?: I2IParams;
  inpaintModel?: NAIInpaintModel;
  inpaintStrength?: number;
  account?: AccountSummary;
  image?: Pick<WorkingImage, "width" | "height"> | null;
  upscaleScale?: UpscaleScale;
  directorTool?: DirectorTool;
  alreadyEncodedVibes?: number;
}): AnlasQuoteResult {
  if (feature === "upscale") {
    return calculateUpscaleAnlas({ image, account, scale: upscaleScale });
  }
  if (feature === "director") {
    return calculateDirectorAnlas({ tool: directorTool, account });
  }
  if (!params) {
    return { ok: false, source: "unavailable", message: "Missing generation parameters for Anlas quote." };
  }
  if (feature === "i2i") {
    return calculateImageGenerationAnlas({
      params,
      account,
      extras,
      alreadyEncodedVibes,
      batchCount: 1,
      action: "img2img",
      strength: i2iParams?.strength ?? 1,
    });
  }
  if (feature === "inpaint") {
    const model = (inpaintModel?.replace(/-inpainting$/, "") || params.model) as NAIModel;
    const inpaintParams = {
      ...params,
      model,
      width: image?.width ? Math.max(64, Math.ceil(image.width / 64) * 64) : params.width,
      height: image?.height ? Math.max(64, Math.ceil(image.height / 64) * 64) : params.height,
    };
    return calculateImageGenerationAnlas({
      params: inpaintParams,
      account,
      extras: { vibeImages: [], charCaptions: [] },
      batchCount: 1,
      action: "infill",
      strength: inpaintStrength ?? 1,
    });
  }
  return calculateImageGenerationAnlas({ params, account, extras, batchCount, action: "generate", alreadyEncodedVibes });
}

File: mobile/lib/billing/anlas.dart
import 'dart:math';

import '../i18n/runtime_text.dart';
import '../models/nai_models.dart';

const _basePixelCoefficient = 2951823174884865e-21;
const _stepPixelCoefficient = 5753298233447344e-22;
const _opusFreeMaxPixels = 1024 * 1024;

enum AnlasQuoteSource { officialApi, estimateFormula, unavailable }

class AnlasQuote {
  final bool ok;
  final int? amount;
  final AnlasQuoteSource source;
  final int? balance;
  final bool insufficient;
  final String message;
  final List<String> details;

  const AnlasQuote({
    required this.ok,
    required this.source,
    required this.message,
    this.amount,
    this.balance,
    this.insufficient = false,
    this.details = const [],
  });

  AnlasQuote asOfficial(
    int officialAmount, {
    int samples = 1,
    Object? language,
  }) {
    final total = (max(0, officialAmount) * max(1, samples)).toInt();
    return AnlasQuote(
      ok: true,
      amount: total,
      source: AnlasQuoteSource.officialApi,
      balance: balance,
      insufficient: balance != null && total > balance!,
      message: _af(language, 'anlas.officialQuote', {'total': total}),
      details: [
        _af(language, 'anlas.officialPerSample', {'amount': officialAmount}),
        if (samples > 1)
          _af(language, 'anlas.officialSamples', {
            'amount': officialAmount,
            'samples': samples,
            'total': total,
          }),
      ],
    );
  }
}

Object _lang(Object? language) => language ?? 'en-US';
String _at(Object? language, String key) =>
    runtimeTextFor(_lang(language), key);
String _af(Object? language, String key, Map<String, Object?> values) =>
    runtimeFormatFor(_lang(language), key, values);

AnlasQuote calculateImageGenerationAnlas({
  required GenerateParams params,
  AccountSummary? account,
  GenerateExtras? extras,
  int batchCount = 1,
  bool imageToImage = false,
  double strength = 1,
  bool forcePaid = false,
  int alreadyEncodedVibes = 0,
  int preciseReferenceCount = 0,
  Object? language,
}) {
  final samples = max(1, batchCount.floor());
  final width = max(1, params.width);
  final height = max(1, params.height);
  final pixels = max(width * height, 65536);
  final steps = max(1, params.steps);
  final normalizedStrength = imageToImage ? strength.clamp(0, 1) : 1.0;
  final v4Plus = params.model.contains('-4');
  final vibeCount = extras?.vibeImages.length ?? 0;
  final details = <String>[];

  final activeOpus =
      account?.hasActiveSubscription == true && (account?.tierLevel ?? 0) >= 3;
  final opusFree = !forcePaid &&
      !imageToImage &&
      activeOpus &&
      pixels <= _opusFreeMaxPixels &&
      steps <= 28;

  var basePerSample = 0;
  if (opusFree) {
    details.add(_at(language, 'anlas.opusFree'));
  } else {
    final smeaMultiplier = !v4Plus && params.smeaDyn
        ? 1.4
        : !v4Plus && params.smea
            ? 1.2
            : 1.0;
    final officialBase = (_basePixelCoefficient * pixels +
            _stepPixelCoefficient * pixels * steps)
        .ceil();
    basePerSample = min(
      140,
      max(2, (officialBase * smeaMultiplier * normalizedStrength).ceil()),
    );
    details.add(
      _af(language, 'anlas.baseCost', {'amount': basePerSample}),
    );
  }

  var total = basePerSample * samples;
  if (v4Plus && vibeCount > 0) {
    final cached = min(vibeCount, max(0, alreadyEncodedVibes));
    final toEncode = max(0, vibeCount - cached);
    final encodeCost = toEncode * 2;
    total += encodeCost;
    details.add(_af(language, 'anlas.vibeEncode', {
      'count': toEncode,
      'amount': encodeCost,
    }));
    if (vibeCount > 4) {
      final extra = 2 * (vibeCount - 4) * samples;
      total += extra;
      details.add(_af(language, 'anlas.vibeExtra', {'amount': extra}));
    }
  }

  if (v4Plus && params.model.contains('4-5') && preciseReferenceCount > 0) {
    final preciseCost = 5 * samples;
    total += preciseCost;
    details.add(_af(language, 'anlas.preciseCost', {
      'samples': samples,
      'amount': preciseCost,
    }));
  }

  final amount = max(0, total.ceil());
  final balance = account?.anlasBalance;
  final insufficient = balance != null && amount > balance;
  return AnlasQuote(
    ok: true,
    amount: amount,
    source: AnlasQuoteSource.estimateFormula,
    balance: balance,
    insufficient: insufficient,
    message: insufficient
        ? _af(language, 'anlas.requiredBalance', {
            'amount': amount,
            'balance': balance,
          })
        : _af(language, 'anlas.estimated', {'amount': amount}),
    details: details,
  );
}

int? extractOfficialAnlasPrice(Object? data) {
  if (data is num && data.isFinite) return max(0, data.ceil());
  if (data is String) {
    final parsed = num.tryParse(data);
    if (parsed != null && parsed.isFinite) return max(0, parsed.ceil());
  }
  if (data is! Map) return null;
  const directKeys = [
    'price',
    'cost',
    'amount',
    'anlas',
    'requestPrice',
    'trainingSteps',
    'trainingStepsCost',
  ];
  for (final key in directKeys) {
    final value = extractOfficialAnlasPrice(data[key]);
    if (value != null) return value;
  }
  for (final key in const ['data', 'result', 'subscription']) {
    final value = extractOfficialAnlasPrice(data[key]);
    if (value != null) return value;
  }
  return null;
}

AnlasQuote calculateInpaintAnlas({
  required GenerateParams params,
  required AccountSummary account,
  required WorkingImage? image,
  required String inpaintModel,
  double strength = 1,
  Object? language,
}) {
  if (image == null || image.width <= 0 || image.height <= 0) {
    return AnlasQuote(
      ok: false,
      source: AnlasQuoteSource.unavailable,
      message: _at(language, 'anlas.loadInpaintImage'),
    );
  }
  final quoteParams = params.copy()
    ..model = inpaintModel.replaceFirst(RegExp(r'-inpainting$'), '')
    ..width = max(64, (image.width / 64).ceil() * 64)
    ..height = max(64, (image.height / 64).ceil() * 64);
  return calculateImageGenerationAnlas(
    params: quoteParams,
    account: account,
    extras: GenerateExtras(),
    imageToImage: true,
    strength: strength,
    language: language,
  );
}

AnlasQuote calculateUpscaleAnlas({
  required WorkingImage? image,
  required AccountSummary account,
  Object? language,
}) {
  if (image == null || image.width <= 0 || image.height <= 0) {
    return AnlasQuote(
      ok: false,
      source: AnlasQuoteSource.unavailable,
      message: _at(language, 'anlas.loadUpscaleImage'),
    );
  }
  const maxPixels = 1024 * 1024;
  final originalPixels = image.width * image.height;
  final ratio =
      originalPixels <= maxPixels ? 1.0 : sqrt(maxPixels / originalPixels);
  final width = max(1, (image.width * ratio).floor());
  final height = max(1, (image.height * ratio).floor());
  final pixels = width * height;
  final details = <String>[
    if (ratio < 1)
      _af(language, 'anlas.upscaleResize', {
        'from': '${image.width}x${image.height}',
        'to': '${width}x$height',
      })
    else
      _af(language, 'anlas.upscaleInput', {
        'size': '${image.width}x${image.height}',
      }),
  ];
  final activeOpus =
      account.hasActiveSubscription == true && (account.tierLevel ?? 0) >= 3;
  var amount = -1;
  if (activeOpus && pixels <= 409600) {
    amount = 0;
  } else if (pixels <= 262144) {
    amount = 1;
  } else if (pixels <= 409600) {
    amount = 2;
  } else if (pixels <= 524288) {
    amount = 3;
  } else if (pixels <= 786432) {
    amount = 5;
  } else if (pixels <= 1048576) {
    amount = 7;
  }
  if (amount < 0) {
    return AnlasQuote(
      ok: false,
      source: AnlasQuoteSource.unavailable,
      balance: account.anlasBalance,
      message: _at(language, 'anlas.upscaleTooLarge'),
      details: details,
    );
  }
  final balance = account.anlasBalance;
  return AnlasQuote(
    ok: true,
    amount: amount,
    source: AnlasQuoteSource.estimateFormula,
    balance: balance,
    insufficient: balance != null && amount > balance,
    message: _af(language, 'anlas.upscaleEstimated', {'amount': amount}),
    details: details,
  );
}

AnlasQuote calculateDirectorAnlas({
  required String tool,
  required AccountSummary account,
  Object? language,
}) {
  final amount = tool == 'bg-removal' ? 65 : 0;
  final balance = account.anlasBalance;
  return AnlasQuote(
    ok: true,
    amount: amount,
    source: AnlasQuoteSource.estimateFormula,
    balance: balance,
    insufficient: balance != null && amount > balance,
    message: amount == 0
        ? _at(language, 'anlas.directorFree')
        : _at(language, 'anlas.directorBgRemoval'),
    details: [
      amount == 0
          ? _at(language, 'anlas.directorFreeDetail')
          : _at(language, 'anlas.directorBgRemovalDetail'),
    ],
  );
}
